File: src/main/java/io/github/regenerativep/compasstracker/CommandListener.java
package io.github.regenerativep.compasstracker;

import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;

import io.github.regenerativep.commandmanager.CommandArgumentType;
import io.github.regenerativep.commandmanager.CommandFunction;
import io.github.regenerativep.commandmanager.CommandManager;
import io.github.regenerativep.commandmanager.CommandSpecifier;

/**
 * provides the interface between command input and the application
 */
public class CommandListener implements CommandExecutor
{
  /**
   * permission for adding anyone as a listener
   */
  public final String PERM_ADD_LISTENER_ANY = "ctrack.addlistener";
  /**
   * permission for adding theirself as a listener
   */
  public final String PERM_ADD_LISTENER_SELF = "ctrack.addlistener.self";
  /**
   * permission for creating and listening to a location target
   */
  public final String PERM_ADD_TARGET_LOCATION = "ctrack.addtargetlocation";
  /**
   * permission for removing any listener from a target
   */
  public final String PERM_REMOVE_LISTENER_ANY = "ctrack.removelistener";
  /**
   * permission for removing theirself from a target
   */
  public final String PERM_REMOVE_LISTENER_SELF = "ctrack.removelistener.self";
  /**
   * permission for adding anyone as a target
   */
  public final String PERM_ADD_TARGET_ANY = "ctrack.addtarget";
  /**
   * permission for adding theirself as a target
   */
  public final String PERM_ADD_TARGET_SELF = "ctrack.addtarget.self";
  /**
   * permission for removing a target
   */
  public final String PERM_REMOVE_TARGET = "ctrack.removetarget";
  /**
   * premission for checking what they are targeting
   */
  public final String PERM_CHECK_TARGET = "ctrack.checktarget";
  /**
   * message for when the command sender is not a player when they should be
   */
  public final String NOT_A_PLAYER = "You are not a player!";
  /**
   * message for when the player cannot be found
   */
  public final String CANNOT_FIND_PLAYER = "Could not find player by the name of \"**name**\".";
  /**
   * message for when the target cannot be found
   */
  public final String CANNOT_FIND_TARGET = "Could not find target by the name of \"**name**\". Do you need to create a target?";
  /**
   * message for when the caller has insufficient permission
   */
  public final String NO_PERMISSION = "Insufficient permissions.";
  /**
   * message for when a target already exists
   */
  public final String TARGET_ALREADY_EXISTS = "Target \"**name**\" already exists.";
  /**
   * message for when the sender is not listening to a target
   */
  public final String NOT_LISTENING = "You are not listening to any targets.";
  /**
   * message for when the sender is listening to a target and to what target
   */
  public final String LISTENING_TO = "You are listening to **name**.";
  /**
   * message for when the sender is not listening to a target
   */
  public final String NO_TARGET = "You are not listening to a target.";
  public final String AUTOLISTEN_STATUS = "Autolistener is **status**.";

  /**
   * a reference to the main application
   */
  private App app;
  /**
   * the command manager
   */
  private CommandManager cmdManager;
  /**
   * constructs a command listener for the compass tracker plugin
   * @param app a reference to the main application
   */
  public CommandListener(App app)
  {
    this.app = app;
    cmdManager = new CommandManager(app);
    /*
    listen { ([x: int] [z: int] | [player]) { [player] } }
    unlisten { [player] }
    relisten { [player] }
    autolisten { [boolean] }
    target { [player] }
    removetarget { [player] }
    who { [player} }
    */
    cmdManager.commands.add(new CommandSpecifier( //ctr listen
      new Object[] { "listen" },
      new CommandArgumentType[] {
        CommandArgumentType.STRING
      },
      new CommandFunction()
      {
        @Override
        public boolean call(CommandSender sender, Object[] args) {
          return tryListen(sender, "");
        }
      },
      new String[] { PERM_ADD_TARGET_LOCATION }
    ));
    cmdManager.commands.add(new CommandSpecifier( //ctr listen [x] [z]
      new Object[] { "listen" },
      new CommandArgumentType[] {
        CommandArgumentType.STRING,
        CommandArgumentType.INTEGER,
        CommandArgumentType.INTEGER
      },
      new CommandFunction()
      {
        @Override
        public boolean call(CommandSender sender, Object[] args) {
          return tryListen((int)args[1], (int)args[2], sender, "");
        }
      },
      new String[] { PERM_ADD_TARGET_LOCATION }
    ));
    cmdManager.commands.add(new CommandSpecifier( //ctr listen [x] [z] [listener]
      new Object[] { "listen" },
      new CommandArgumentType[] {
        CommandArgumentType.STRING,
        CommandArgumentType.INTEGER,
        CommandArgumentType.INTEGER,
        CommandArgumentType.STRING
      },
      new CommandFunction()
      {
        @Override
        public boolean call(CommandSender sender, Object[] args) {
          return tryListen((int)args[1], (int)args[2], sender, (String)args[3]);
        }
      },
      new String[] { PERM_ADD_LISTENER_ANY }
    ));
    cmdManager.commands.add(new CommandSpecifier( //ctr listen [target name]
      new Object[] { "listen" },
      new CommandArgumentType[] {
        CommandArgumentType.STRING,
        CommandArgumentType.STRING
      },
      new CommandFunction()
      {
        @Override
        public boolean call(CommandSender sender, Object[] args) {
          return tryListen((String)args[1], sender, "");
        }
      },
      new String[] { PERM_ADD_LISTENER_SELF }
    ));
    cmdManager.commands.add(new CommandSpecifier( //ctr listen [target name] [listener name]
      new Object[] { "listen" },
      new CommandArgumentType[] {
        CommandArgumentType.STRING,
        CommandArgumentType.STRING,
        CommandArgumentType.STRING
      },
      new CommandFunction()
      {
        @Override
        public boolean call(CommandSender sender, Object[] args) {
          return tryListen((String)args[1], sender, (String)args[2]);
        }
      },
      new String[] { PERM_ADD_LISTENER_ANY }
    ));
    cmdManager.commands.add(new CommandSpecifier( //ctr remove
      new Object[] { "unlisten" },
      new CommandArgumentType[] {
        CommandArgumentType.STRING
      },
      new CommandFunction()
      {
        @Override
        public boolean call(CommandSender sender, Object[] args) {
          return tryUnlisten(sender, "");
        }
      },
      new String[] { PERM_ADD_LISTENER_SELF }
    ));
    cmdManager.commands.add(new CommandSpecifier( //ctr remove [listener name]
      new Object[] { "unlisten" },
      new CommandArgumentType[] {
        CommandArgumentType.STRING,
        CommandArgumentType.STRING
      },
      new CommandFunction()
      {
        @Override
        public boolean call(CommandSender sender, Object[] args) {
          return tryUnlisten(sender, (String)args[1]);
        }
      },
      new String[] { PERM_ADD_LISTENER_ANY }
    ));
    cmdManager.commands.add(new CommandSpecifier( //ctr remove
      new Object[] { "relisten" },
      new CommandArgumentType[] {
        CommandArgumentType.STRING
      },
      new CommandFunction()
      {
        @Override
        public boolean call(CommandSender sender, Object[] args) {
          return tryRelisten(sender, "");
        }
      },
      new String[] { PERM_ADD_LISTENER_SELF }
    ));
    cmdManager.commands.add(new CommandSpecifier( //ctr remove [listener name]
      new Object[] { "relisten" },
      new CommandArgumentType[] {
        CommandArgumentType.STRING,
        CommandArgumentType.STRING
      },
      new CommandFunction()
      {
        @Override
        public boolean call(CommandSender sender, Object[] args) {
          return tryRelisten(sender, (String)args[1]);
        }
      },
      new String[] { PERM_ADD_LISTENER_ANY }
    ));
    cmdManager.commands.add(new CommandSpecifier( //ctr autolisten
      new Object[] { "autolisten" },
      new CommandArgumentType[] {
        CommandArgumentType.STRING
      },
      new CommandFunction()
      {
        @Override
        public boolean call(CommandSender sender, Object[] args) {
          return tryAutoListen(sender);
        }
      },
      new String[] { PERM_ADD_LISTENER_ANY }
    ));
    cmdManager.commands.add(new CommandSpecifier( //ctr autolisten [boolean]
      new Object[] { "autolisten" },
      new CommandArgumentType[] {
        CommandArgumentType.STRING,
        CommandArgumentType.BOOLEAN
      },
      new CommandFunction()
      {
        @Override
        public boolean call(CommandSender sender, Object[] args) {
          return tryAutoListen(sender, (boolean)args[1]);
        }
      },
      new String[] { PERM_ADD_LISTENER_ANY }
    ));
    cmdManager.commands.add(new CommandSpecifier( //ctr target
      new Object[] { "target" },
      new CommandArgumentType[] {
        CommandArgumentType.STRING
      },
      new CommandFunction()
      {
        @Override
        public boolean call(CommandSender sender, Object[] args) {
          return tryTarget(sender, "");
        }
      },
      new String[] { PERM_ADD_TARGET_SELF }
    ));
    cmdManager.commands.add(new CommandSpecifier( //ctr target [target name]
      new Object[] { "target" },
      new CommandArgumentType[] {
        CommandArgumentType.STRING,
        CommandArgumentType.STRING
      },
      new CommandFunction()
      {
        @Override
        public boolean call(CommandSender sender, Object[] args) {
          return tryTarget(sender, (String)args[1]);
        }
      },
      new String[] { PERM_ADD_TARGET_ANY }
    ));
    cmdManager.commands.add(new CommandSpecifier( //ctr stop
      new Object[] { "removetarget" },
      new CommandArgumentType[] {
        CommandArgumentType.STRING
      },
      new CommandFunction()
      {
        @Override
        public boolean call(CommandSender sender, Object[] args) {
          return tryRemoveTarget(sender, "");
        }
      },
      new String[] { PERM_REMOVE_TARGET }
    ));
    cmdManager.commands.add(new CommandSpecifier( //ctr stop [target name]
      new Object[] { "removetarget" },
      new CommandArgumentType[] {
        CommandArgumentType.STRING,
        CommandArgumentType.STRING
      },
      new CommandFunction()
      {
        @Override
        public boolean call(CommandSender sender, Object[] args) {
          return tryRemoveTarget(sender, (String)args[1]);
        }
      },
      new String[] { PERM_REMOVE_TARGET }
    ));
    cmdManager.commands.add(new CommandSpecifier( //ctr who
      new Object[] { "who" },
      new CommandArgumentType[] {
        CommandArgumentType.STRING
      },
      new CommandFunction()
      {
        @Override
        public boolean call(CommandSender sender, Object[] args) {
          return tryCheck(sender, "");
        }
      },
      new String[] { }
    ));
    cmdManager.commands.add(new CommandSpecifier( //ctr who [listener name]
      new Object[] { "who" },
      new CommandArgumentType[] {
        CommandArgumentType.STRING,
        CommandArgumentType.STRING
      },
      new CommandFunction()
      {
        @Override
        public boolean call(CommandSender sender, Object[] args) {
          return tryCheck(sender, (String)args[1]);
        }
      },
      new String[] { PERM_CHECK_TARGET }
    ));
  }
  @Override
  public boolean onCommand(CommandSender sender, Command command, String label, String[] args)
  {
    return cmdManager.inputCommand(sender, args);
  }
  /**
   * called to display not a player error
   * @param sender sender to send message to
   */
  public void errorNotAPlayer(CommandSender sender)
  {
    sender.sendMessage(NOT_A_PLAYER);
  }
  /**
   * called to display cannot find player error
   * @param sender sender to send message to
   * @param playerName name of the player
   */
  public void errorCannotFindPlayer(CommandSender sender, String playerName)
  {
    sender.sendMessage(CANNOT_FIND_PLAYER.replace("**name**", playerName));
  }
  /**
   * called to display cannot find target error
   * @param sender sender to send message to
   * @param playerName name of the player
   */
  public void errorCannotFindTarget(CommandSender sender, String playerName)
  {
    sender.sendMessage(CANNOT_FIND_TARGET.replace("**name**", playerName));
  }
  /**
   * called to display no permissions error
   * @param sender sender to send message to
   */
  public void errorNoPermission(CommandSender sender)
  {
    sender.sendMessage(NO_PERMISSION);
  }
  /**
   * called to display target already exists error
   * @param sender sender to send message to
   * @param targetName name of the target
   */
  public void errorTargetAlreadyExists(CommandSender sender, String targetName)
  {
    sender.sendMessage(TARGET_ALREADY_EXISTS.replace("**name**", targetName));
  }
  /**
   * gets a player for a command given their name
   * @param sender sender to use if no listener specified
   * @param listenerName name of the listener
   * @return the player found; null if not found
   */
  public Player getPlayerFromName(CommandSender sender, String listenerName)
  {
    Player player;
    if(listenerName.length() == 0)
    {
      if(sender instanceof Player)
      {
        player = (Player)sender;
      }
      else
      {
        errorNotAPlayer(sender);
        return null;
      }
    }
    else
    {
      player = app.getServer().getPlayerExact(listenerName);
      if(player == null)
      {
        errorCannotFindPlayer(sender, listenerName);
        return null;
      }
    }
    return player;
  }
  /**
   * tries to listen given inputs
   * @param x x position
   * @param z z position
   * @param sender command sender
   * @param listenerName name of listener
   * @return if successful
   */
  public boolean tryListen(int x, int z, CommandSender sender, String listenerName)
  {
    Player player = getPlayerFromName(sender, listenerName);
    if(player == null) return false;
    Location loc = new Location(player.getWorld(), x, player.getLocation().getY(), z);
    app.listen(player);
    app.setListenerTarget(player.getName(), loc);
    return true;
  }
  /**
   * tries to listen based on given inputs
   * @param targetName name of the target to listen to
   * @param sender command sender
   * @param listenerName name of listener
   * @return if successful
   */
  public boolean tryListen(String targetName, CommandSender sender, String listenerName)
  {
    Player player = getPlayerFromName(sender, listenerName);
    if(player == null) return false;
    boolean targetExists = app.targetExists(targetName);
    if(!targetExists)
    {
      errorCannotFindTarget(sender, targetName);
      return false;
    }
    app.listen(player);
    app.setListenerTarget(player.getName(), targetName);
    return true;
  }
  /**
   * tries to remove based on given inputs
   * @param sender command sender
   * @param listenerName name of listener
   * @return if successful
   */
  public boolean tryRelisten(CommandSender sender, String listenerName)
  {
    Player player = getPlayerFromName(sender, listenerName);
    if(player == null) return false;
    app.unlisten(player.getName());
    app.listen(player);
    return true;
  }
  public boolean tryListen(CommandSender sender, String listenerName)
  {
    Player player = getPlayerFromName(sender, listenerName);
    if(player == null) return false;
    app.listen(player);
    return true;
  }
  public boolean tryUnlisten(CommandSender sender, String listenerName)
  {
    Player player = getPlayerFromName(sender, listenerName);
    if(player == null) return false;
    app.unlisten(player.getName());
    return true;
  }
  /**
   * tries to create a target based on given inputs
   * @param sender command sender
   * @param listenerName name of listener
   * @return if successful
   */
  public boolean tryTarget(CommandSender sender, String targetName)
  {
    Player targetPlayer = getPlayerFromName(sender, targetName);
    if(targetPlayer == null) return false;
    targetName = targetPlayer.getName();
    if(app.targetExists(targetName))
    {
      errorTargetAlreadyExists(sender, targetName);
      return false;
    }
    app.addTarget(targetName);
    return true;
  }
  /**
   * tries to remove a target based on given inputs
   * @param sender command sender
   * @param listenerName name of listener
   * @return if successful
   */
  public boolean tryRemoveTarget(CommandSender sender, String targetName)
  {
    Player targetPlayer = getPlayerFromName(sender, targetName);
    if(targetPlayer == null) return false;
    targetName = targetPlayer.getName();
    if(!app.targetExists(targetName))
    {
      errorCannotFindTarget(sender, targetName);
      return false;
    }
    app.removeTarget(targetName);
    return true;
  }
  public void sendAutoListenStatus(CommandSender sender)
  {
    String status = "";
    if(app.autoListenEnabled())
    {
      status = "enabled";
    }
    else
    {
      status = "disabled";
    }
    sender.sendMessage(AUTOLISTEN_STATUS.replace("**status**", status));
  }
  public boolean tryAutoListen(CommandSender sender)
  {
    sendAutoListenStatus(sender);
    return true;
  }
  public boolean tryAutoListen(CommandSender sender, boolean status)
  {
    app.setAutoListen(status);
    sendAutoListenStatus(sender);
    return true;
  }
  /**
   * checks and send to sender who they are listening to
   * @param sender command sender
   * @param listenerName name of listener
   * @return if successful
   */
  public boolean tryCheck(CommandSender sender, String playerName)
  {
    Player player = getPlayerFromName(sender, playerName);
    if(player == null) return false;
    playerName = player.getName();
    PlayerListener listener = app.playerMap.get(playerName);
    if(listener != null)
    {
      ILocatable target = listener.getTarget();
      if(target == null)
      {
        sender.sendMessage(NO_TARGET);
      }
      sender.sendMessage(LISTENING_TO.replace("**name**", target.getLocationDescription()));
    }
    else
    {
      sender.sendMessage(NOT_LISTENING);
    }
    return true;
  }
}

File: src/main/kotlin/io/github/regenerativep/compasstracker/CommandListener.kt
package io.github.regenerativep.compasstracker;

import io.github.regenerativep.commandmanager.CommandSpecifier;
import io.github.regenerativep.commandmanager.CommandArgumentType;
import io.github.regenerativep.commandmanager.inputCommand;
import io.github.regenerativep.compasstracker.giveCompass;

import org.bukkit.command.CommandSender;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.World;
import org.bukkit.entity.Player;

val PERM_GIVE_SELF = "ctrack.give.self"
val PERM_GIVE_ANY = "ctrack.give"
val PERM_MANAGE_TARGETS = "ctrack.target"
val PERM_MANAGE_AUTOGIVE = "ctrack.autogive"
val PERM_MANAGE_AUTOTARGET = "ctrack.autotarget"
val PERM_MANAGE_ENVIRONMENT = "ctrack.environment"
val PERM_MANAGE_TICKRATE = "ctrack.tickrate"

fun errTargetExists(sender: CommandSender, targetName: String): Boolean {
    sender.sendMessage("Error: Target \"${targetName}\" already exists.")
    return false
}
fun errTargetDoesNotExist(sender: CommandSender, targetName: String): Boolean {
    sender.sendMessage("Error: Target \"${targetName}\" doesn't exist.")
    return false
}
fun errPlayerDoesNotExist(sender: CommandSender, targetName: String): Boolean {
    sender.sendMessage("Error: Player \"${targetName}\" doesn't exist.")
    return false
}
fun errYouAreNotTarget(sender: CommandSender): Boolean {
    sender.sendMessage("Error: You are not a target.")
    return false
}
fun errTargetHasCompass(sender: CommandSender, targetName: String): Boolean {
    sender.sendMessage("Error: Target \"${targetName}\" already has a compass.")
    return false
}
fun errYouHaveCompass(sender: CommandSender): Boolean {
    sender.sendMessage("Error: You already have a compass.")
    return false
}
fun errYouNotPlayer(sender: CommandSender): Boolean {
    sender.sendMessage("Error: You are not a player.")
    return false
}
fun errIndeterminableEnvironment(sender: CommandSender): Boolean {
    sender.sendMessage("Error: Cannot determine your environment.")
    return false
}
fun warnTargetNotInServer(sender: CommandSender, targetName: String): Boolean {
    sender.sendMessage("Warning: Target \"${targetName}\" is not in the server.")
    return false
}
fun sucTargetedPlayer(sender: CommandSender, targetName: String): Boolean {
    sender.sendMessage("Success: Player \"${targetName}\" has been added as a target.")
    return true
}
fun sucRetargetedPlayer(sender: CommandSender, targetName: String): Boolean {
    sender.sendMessage("Success: Player \"${targetName}\" has been re-added as a target.")
    return true
}
fun sucTargetedSelf(sender: CommandSender): Boolean {
    sender.sendMessage("Success: You have been added as a target.")
    return true
}
fun sucRetargetedSelf(sender: CommandSender): Boolean {
    sender.sendMessage("Success: You have been re-added as a target.")
    return true
}
fun sucRemovedTarget(sender: CommandSender, targetName: String): Boolean {
    sender.sendMessage("Success: Player \"${targetName}\" has been removed from the targets.")
    return true
}
fun sucRemovedTargetSelf(sender: CommandSender): Boolean {
    sender.sendMessage("Success: You have been removed from the targets.")
    return true
}
fun sucGaveCompassSelf(sender: CommandSender): Boolean {
    sender.sendMessage("Success: Gave you a compass.")
    return true
}
fun sucGaveCompass(sender: CommandSender, playerName: String): Boolean {
    sender.sendMessage("Success: Gave player \"${playerName}\" a compass.")
    return true
}
fun sucSetEnvironment(sender: CommandSender, environment: World.Environment, status: Boolean): Boolean {
    sender.sendMessage("Success: Set the environment \"${environment}\" to ${status}")
    return true
}
fun sucNowTrackingSelf(sender: CommandSender, targetName: String): Boolean {
    sender.sendMessage("Success: You are now tracking player \"${targetName}\".")
    return true
}
fun sucNowTracking(sender: CommandSender, targetName: String, listenerName: String): Boolean {
    sender.sendMessage("Success: Player \"${listenerName}\" is now tracking player \"${targetName}\".")
    return true
}
fun sucAutoGiveStatus(sender: CommandSender, status: Boolean): Boolean {
    sender.sendMessage("Success: Automatically giving compasses to players has been set to ${status}.")
    return true
}
fun sucChangeTickRate(sender: CommandSender, tickrate: Long): Boolean {
    sender.sendMessage("Success: Set the tick amount between compass updates to ${tickrate}")
    return true
}
fun sucAutoTargetStatus(sender: CommandSender, status: Boolean): Boolean {
    sender.sendMessage("Success: Automatically targeting players as they join has been set to ${status}.")
    return true
}

class CommandListener(val app: CompassTracker, val cmdName: String) : CommandExecutor {
    val commands: List<CommandSpecifier> = listOf(
        CommandSpecifier(
            arrayOf("target"), arrayOf(CommandArgumentType.STRING), { sender, _
                -> if(sender is Player) {
                    if(sender.name in app.targets.keys) {
                        sucRetargetedSelf(sender)
                    }
                    else {
                        sucTargetedSelf(sender)
                    }.let { res ->
                        app.targets.set(sender.name, TargetListener(sender.name))
                        res
                    }
                }
                else {
                    errYouNotPlayer(sender)
                }
            }, arrayOf(PERM_MANAGE_TARGETS)
        ),
        CommandSpecifier(
            arrayOf("target"), arrayOf(CommandArgumentType.STRING, CommandArgumentType.STRING), { sender, args
                -> (args[1] as String).let { targetName
                    -> if(!playerExists(targetName)) {
                        warnTargetNotInServer(sender, targetName)
                    }
                    if(targetName in app.targets.keys) {
                        sucRetargetedPlayer(sender, targetName)
                    }
                    else {
                        sucTargetedPlayer(sender, targetName)
                    }.let{ res ->
                        app.targets.set(targetName, TargetListener(targetName))
                        res
                    }
                }
            }, arrayOf(PERM_MANAGE_TARGETS)
        ),
        CommandSpecifier(
            arrayOf("removetarget"), arrayOf(CommandArgumentType.STRING), { sender, _
                -> if(sender is Player) {
                    if(app.targets.remove(sender.name) == null) {
                        errYouAreNotTarget(sender)
                    }
                    else {
                        sucRemovedTargetSelf(sender)
                    }
                }
                else {
                    errYouNotPlayer(sender)
                }
            }, arrayOf(PERM_MANAGE_TARGETS)
        ),
        CommandSpecifier(
            arrayOf("removetarget"), arrayOf(CommandArgumentType.STRING, CommandArgumentType.STRING), { sender, args
                -> (args[1] as String).let { targetName
                    -> if(app.targets.remove(targetName) == null) {
                        errTargetDoesNotExist(sender, targetName)
                    }
                    else {
                        sucRemovedTarget(sender, targetName)
                    }
                }
            }, arrayOf(PERM_MANAGE_TARGETS)
        ),
        CommandSpecifier(
            arrayOf("targetall"), arrayOf(CommandArgumentType.STRING), { sender, _ ->
                val targetedPlayers = app.server.onlinePlayers.filter { it.name !in app.targets.keys }.map { it -> it.name }
                targetedPlayers.forEach { name ->
                    app.targets.set(name, TargetListener(name))
                }
                sender.sendMessage(
                    if(targetedPlayers.size > 0) {
                        targetedPlayers.reduce { a, b -> a + ", " + b }.let { "Targeted ${it}" }
                    }
                    else {
                        "Did not target anyone new."
                    }
                )
                true
            }, arrayOf(PERM_MANAGE_TARGETS)
        ),
        CommandSpecifier(
            arrayOf("cleartargets"), arrayOf(CommandArgumentType.STRING), { sender, _ ->
                val targetedPlayers = app.targets.keys.toTypedArray()
                targetedPlayers.forEach { name ->
                    app.targets.remove(name)
                }
                sender.sendMessage(
                    if(targetedPlayers.size > 0) {
                        targetedPlayers.reduce { a, b -> a + ", " + b }.let { "Removed ${it} from targets" }
                    }
                    else {
                        "Did not remove any new targets."
                    }
                )
                true
            }, arrayOf(PERM_MANAGE_TARGETS)
        ),
        CommandSpecifier(
            arrayOf("targetlist"), arrayOf(CommandArgumentType.STRING), { sender, _
                -> sender.sendMessage(
                    if(app.targets.size > 0) {
                        app.targets.keys.reduce { a, b -> a + ", " + b }.let { "Targets: ${it}" }
                    }
                    else {
                        "No targets."
                    }
                )
                true
            }
        ),
        CommandSpecifier(
            arrayOf("give"), arrayOf(CommandArgumentType.STRING), { sender, _ ->
                if(sender is Player) {
                    if(!sender.giveCompass(app)) {
                        errYouHaveCompass(sender)
                    }
                    else {
                        sucGaveCompassSelf(sender)
                    }
                }
                else {
                    errYouNotPlayer(sender)
                }
            }, arrayOf(PERM_GIVE_SELF)
        ),
        CommandSpecifier(
            arrayOf("give"), arrayOf(CommandArgumentType.STRING, CommandArgumentType.STRING), { sender, args ->
                (args[1] as String).let { targetName ->
                    app.server.getPlayerExact(targetName)?.let { player ->
                        if(!player.giveCompass(app)) {
                            errTargetHasCompass(sender, targetName)
                        }
                        else {
                            sucGaveCompass(sender, targetName)
                        }
                    } ?: errPlayerDoesNotExist(sender, targetName)
                }
            }, arrayOf(PERM_GIVE_ANY)
        ),
        CommandSpecifier(
            arrayOf("giveall"), arrayOf(CommandArgumentType.STRING), { sender, _ ->
                val playersToGive = this.app.server.onlinePlayers.filter { player -> player.giveCompass(this.app) }.map { it.name }
                sender.sendMessage("Success: " +
                    if(playersToGive.size > 0) {
                        playersToGive.reduce { a, b -> a + ", " + b }.let { "Gave compasses to ${it}" }
                    }
                    else {
                        "Did not give anyone any compasses."
                    }
                )
                true
            }, arrayOf(PERM_GIVE_ANY)
        ),
        CommandSpecifier(
            arrayOf("environment"), arrayOf(CommandArgumentType.STRING, CommandArgumentType.BOOLEAN), { sender, args
                -> (args[1] as Boolean).let { enabled
                    -> if(sender is Player) {
                        sender.location.world?.environment?.let { env
                            -> app.setEnvironment(env, enabled)
                            sucSetEnvironment(sender, env, enabled)
                        }
                        ?: errIndeterminableEnvironment(sender)
                    }
                    else {
                        errYouNotPlayer(sender)
                    }
                }
            }, arrayOf(PERM_MANAGE_ENVIRONMENT)
        ),
        CommandSpecifier(
            arrayOf("environment"), arrayOf(CommandArgumentType.STRING, CommandArgumentType.BOOLEAN, CommandArgumentType.ENVIRONMENT), { sender, args
                -> (args[1] as Boolean).let { enabled
                    -> (args[2] as World.Environment).let { env
                        -> app.setEnvironment(env, enabled)
                        sucSetEnvironment(sender, env, enabled)
                    }
                }
            }, arrayOf(PERM_MANAGE_ENVIRONMENT)
        ),
        CommandSpecifier(
            arrayOf("environment"), arrayOf(CommandArgumentType.STRING, CommandArgumentType.ENVIRONMENT, CommandArgumentType.BOOLEAN), { sender, args
                -> (args[2] as Boolean).let { enabled
                    -> (args[1] as World.Environment).let { env
                        -> app.setEnvironment(env, enabled)
                        sucSetEnvironment(sender, env, enabled)
                    }
                }
            }, arrayOf(PERM_MANAGE_ENVIRONMENT)
        ),
        CommandSpecifier(
            arrayOf("environmentlist"), arrayOf(CommandArgumentType.STRING), { sender, _
                -> sender.sendMessage(
                    if(app.permittedEnvironments.size > 0) {
                        app.permittedEnvironments.map { " ${it}" }.reduce { a, b -> a + b }.let { "Trackable environments: ${it}" }
                    }
                    else {
                        "No trackable environments."
                    }
                )
                true
            }
        ),
        CommandSpecifier(
            arrayOf("who"), arrayOf(CommandArgumentType.STRING), { sender, _
                -> if(sender is Player) {
                    app.listeners.get(sender.name)?.targetName?.let { targetName
                        -> sender.sendMessage("You are tracking \"${targetName}\".")
                    }
                    ?: sender.sendMessage("You are not tracking anyone.")
                    true
                }
                else {
                    errYouNotPlayer(sender)
                }
            }
        ),
        CommandSpecifier(
            arrayOf("who"), arrayOf(CommandArgumentType.STRING, CommandArgumentType.STRING), { sender, args
                -> (args[1] as String).let { listenerName
                    -> app.listeners.get(listenerName)?.targetName?.let { targetName
                        -> sender.sendMessage("Player \"${listenerName}\" is tracking \"${targetName}\".")
                    }
                    ?: sender.sendMessage("Player \"${listenerName}\" is not tracking anyone.")
                    true
                }
            }
        ),
        CommandSpecifier(
            arrayOf("track"), arrayOf(CommandArgumentType.STRING, CommandArgumentType.STRING), { sender, args
                -> (args[1] as String).let { targetName
                    -> if(sender is Player) {
                        if(targetName in app.targets.keys) {
                            app.setTarget(sender.name, targetName)
                            sucNowTrackingSelf(sender, targetName)
                        }
                        else {
                            errTargetDoesNotExist(sender, targetName)
                        }
                    }
                    else {
                        errTargetDoesNotExist(sender, targetName)
                    }
                }
            }
        ),
        CommandSpecifier(
            arrayOf("track"), arrayOf(CommandArgumentType.STRING, CommandArgumentType.STRING, CommandArgumentType.STRING), { sender, args
                -> (args[1] as String).let { targetName
                    -> (args[2] as String).let { listenerName
                        -> if(app.server.getPlayerExact(listenerName) != null) {
                            if(targetName in app.targets.keys) {
                                app.setTarget(listenerName, targetName)
                                sucNowTracking(sender, targetName, listenerName)
                            }
                            else {
                                errTargetDoesNotExist(sender, targetName)
                            }
                        }
                        else {
                            errPlayerDoesNotExist(sender, listenerName)
                        }
                    }
                }
            }, arrayOf(PERM_MANAGE_TARGETS)
        ),
        CommandSpecifier(
            arrayOf("autogive"), arrayOf(CommandArgumentType.STRING, CommandArgumentType.BOOLEAN), { sender, args
                -> (args[1] as Boolean).let { enabled
                    -> if(enabled) {
                        app.server.onlinePlayers.forEach { it.giveCompass(app) }
                    }
                    app.autoGiveCompass = enabled
                    sucAutoGiveStatus(sender, enabled)
                }
            }, arrayOf(PERM_MANAGE_AUTOGIVE)
        ),
        CommandSpecifier(
            arrayOf("tickrate"), arrayOf(CommandArgumentType.STRING, CommandArgumentType.INTEGER), { sender, args
                -> (args[1] as Int).toLong().let { ticks
                    -> app.runUpdateTimer(ticks)
                    sucChangeTickRate(sender, ticks)
                }
            }, arrayOf(PERM_MANAGE_TICKRATE)
        ),
        CommandSpecifier(
            arrayOf("autotarget"), arrayOf(CommandArgumentType.STRING, CommandArgumentType.BOOLEAN), { sender, args
                -> (args[1] as Boolean).let { enabled
                    -> app.autoTarget = enabled
                    sucAutoTargetStatus(sender, enabled)
                }
            }, arrayOf(PERM_MANAGE_AUTOTARGET)
        )
    )
    override fun onCommand(sender: CommandSender, command: Command, label: String, args: Array<String>): Boolean {
        return inputCommand(this.commands, sender, args, this.cmdName)
    }
    fun playerExists(targetName: String): Boolean {
        return this.app.server.getPlayerExact(targetName) != null
    }
}